Apml i3c

mahkurap opened the pull request:
drivers : i3c : add support for APML i3c bus

Following changes are required to support APML I3C bus on Venice based platforms:
1. configure internal pullups in the device tree
2. set the NACK retry count to max 
3. set the dynamic address with the assigned address for the device allocation to be successful in the device address table

verified : tested on Morocco

supven01 commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -745,6 +745,12 @@ static int i3c_hci_attach_i3c_dev(struct i3c_dev_desc *dev)
+
+#ifdef CONFIG_ARCH_ASPEED
+		/* set the DEV_NACK_RETRY_CNT to max for this device */
+		mipi_i3c_hci_dat_v1.set_flags(hci, ret, W0_MASK(30, 29), 0);
```
Are there no #define for 30 and 29?

mahkurap replied:
moved to dat_v1.c::alloc_entry() as a default flag after allocating an entry for this device.

supven01 commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -201,12 +201,14 @@ static void aspeed_i3c_of_populate_bus_timing(struct i3c_hci *hci, struct device
 	/* core_period is in nanosecond */
 	core_period = DIV_ROUND_UP(1000000000, core_rate);
 
+#if 0
```
Please remove this block

supven01 commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -471,10 +473,12 @@ static int i3c_hci_send_ccc_cmd(struct i3c_master_controller *m,
 	    ccc->rnw, ccc->dbp, ccc->db, ccc->ndests,
 	    ccc->dests[0].payload.len);
 
+#if 0
```
Please remove this block. History will be available in GitHub anyway

mahkurap committed:
drivers : mipi_hci_i3c : do not skip CCC for APML I3C

Handling of the mipi_i3c_hci module input parameter i3c_device_power
returns success when the CCC commands are skipped. This causes
kernel panic when the device node is created for the driver. Since
all the address assignment commands are returned success, driver
tries to create the device file with the data from dynamic address
which is not populated, resulting an attempt to create a duplicate
device file. This is causing kernel panic. To fix this, since we
have the i3c mux only on SPD I3C bus which is JESD compliant, a
check is added not to skip the CCC commands.

Tested : verified on Congo and Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
drivers : mipi_hci_i3c : set DEV_NACK_RETRY_CNT to three

Configures the Device Nack Retry Count to max retry value of three
in the Device Address Table (DAT) for each slave device configured/
detected on the I3C bus. This register setting is specific to the
ASPEED master controller.

Tested: verified on Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
ARM64 : dts : aspeed : Kenya : fix apml bus config

Do the following changes to apml i3c4 bus configuration:
- Add internal-pullup tag
- update the i3c clock frequency to 12.5MHz
- remove the JESD403 bus context

Tested: verified on Kenya

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
ARM64 : dts : aspeed : Nigeria : fix apml bus config

Do the following changes to apml i3c4 bus configuration:
- Add internal-pullup tag
- update the i3c clock frequency to 12.5MHz
- remove the JESD403 bus context

Add i3c5 bus configuration similar to i3c4 except the static
addresses for rmi and tsi devices.

Tested: verified on Nigeria

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

ojayanth commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -203,7 +203,7 @@ static void aspeed_i3c_of_populate_bus_timing(struct i3c_hci *hci, struct device
 	/* Workaround . Need to remove once hardware support is available
 	   for internal LDO powerup support is available. */
-	if (!i3c_device_power) {
+	if (!i3c_device_power && hci->master.bus.context == I3C_BUS_CONTEXT_JESD403) {
```
since this workaround , can you add note saying this also should reverted as part of workaround removal.

mahkurap replied:
> since this workaround , can you add note saying this also should reverted as part of workaround removal.

will do

ojayanth commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -471,7 +471,7 @@ static int i3c_hci_send_ccc_cmd(struct i3c_master_controller *m,
 	    ccc->dests[0].payload.len);
 
-	if (!i3c_device_power) {
+	if (!i3c_device_power && m->bus.context == I3C_BUS_CONTEXT_JESD403) {
```
since this workaround , can you add note saying this also should reverted as part of workaround removal.

mahkurap replied:
> since this workaround , can you add note saying this also should reverted as part of workaround removal.

I can.

ojayanth commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-congo.dts:
```diff
@@ -449,8 +449,8 @@
 &i3c4 {
 	status = "okay";
 	initial-role = "primary";
```
Please split to machine specific commits , that is the process we follows.

This commit also changes frequency , please update commit message with this details

ojayanth commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-morocco.dts:
```diff
@@ -619,6 +620,24 @@
 		assigned-address = <0x3c>;
```
Looks like commit message is just a copy of previous commit , Not specific to this commit?

ojayanth commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-kenya.dts:
```diff
@@ -319,9 +319,9 @@
-        bus-context = /bits/ 8 <I3C_BUS_CONTEXT_JESD403>;
-        i3c-scl-hz = <10000000>;
-        i2c-scl-hz = <1000000>;
+	internal-pullup = <2>;
```
This commit also changes frequency , please update commit message with this detAils

ojayanth replied:
Looks like missing Bus-context tag removal related updated in the commit message.

ojayanth commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-nigeria.dts:
```diff
@@ -386,9 +386,10 @@
 	initial-role = "primary";
-	bus-context = /bits/ 8 <I3C_BUS_CONTEXT_JESD403>;
-	i3c-scl-hz = <6000000>;
+	internal-pullup = <2>;
```
Commit message update is required based on the changes you added here.

ojayanth replied:
Looks like missing bus-context tag removal in the commit message.

mahkurap replied:
> Looks like missing bus-context tag removal in the commit message.

Done.

ojayanth commented on drivers/i3c/master.c:
```diff
@@ -1722,6 +1722,7 @@ static int i3c_master_early_i3c_dev_add(struct i3c_master_controller *master,
 {
 	struct i3c_device_info info = {
 		.static_addr = boardinfo->static_addr,
+		.dyn_addr = boardinfo->init_dyn_addr,
```
Billy's comment: 
The dynamic address will be assigned only after the SETDASA succeeds; until then, the device should only respond to its static address.
https://github.com/AMDESE/linux-aspeed/blob/integ_sp7/drivers/i3c/master.c#L1740-L1745

mahkurap replied:
> Billy's comment: The dynamic address will be assigned only after the SETDASA succeeds; until then, the device should only respond to its static address. https://github.com/AMDESE/linux-aspeed/blob/integ_sp7/drivers/i3c/master.c#L1740-L1745

@ojayanth  - Following check fails when hci_dat_v1_alloc_entry() is called with zero dynamic address. 
hci_dat_v1_alloc_entry
https://github.com/AMDESE/linux-aspeed/blob/integ_sp7/drivers/i3c/master/mipi-i3c-hci/dat_v1.c#L100C1-L101C18

For SBRMI/SBTSI devices, driver need to send the SETDASA command as part of the early init process during the i3c bus registration. Before sending the SETDASA command driver reserves an entry in the device address table with dynamic address as the index which fails with out my fix and subsequently SETDASA command will not be sent, and the device nodes will not be created.  Please check with Billy what could be a suggested fix. Below is the call stack:

i3c_master_register()
i3c_master_bus_init()
i3c_master_early_i3c_dev_add() 
i3c_hci_attach_i3c_dev() ==>Here instead of using the info.dyn_addr as parameter, we can take the dynamic address mentioned in the device tree. this will avoid change to the toplevel driver (master.c). 
hci_dat_v1_alloc_entry()

mahkurap committed:
ARM64 : dts : aspeed : Congo : fix apml bus config

Add internal-pullup tag, update the i3c clock frequency to 12.5MHz,
and remove the JESD403 compliance tag for i3c4.

tested: verified on Congo

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

mahkurap committed:
ARM64 : dts : aspeed : Morocco : fix apml bus config

Add internal-pullup tag, remove JESD403 compliance tag, and update
the i3c clock frequency to 12.5MHz for i3c4.

Add i3c5 bus configuration similar to i3c4 except the static
addresses for rmi and tsi devices.

Tested: verified on Morocco

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>

billy-tsai committed:
i3c: mipi-i3c-hci: aspeed : Support SETDASA CCC

When the I3C subsystem wants to assign a dynamic address using the SETDASA
CCC, it needs to attach the I3C device with device info that includes only
the static address. In the HCI, if the driver want to send this SETDASA
CCC, a DAT entry is required to temporarily fill the device’s static
address into the dynamic address field. Afterward, the reattach API will
be executed to update the DAT with the correct dynamic addrees value.

Signed-off-by: Billy Tsai <billy_tsai@aspeedtech.com>
Change-Id: If0576d46210daef7e4572834462aa52b3d79bef1

billy-tsai committed:
i3c: mipi-i3c-hci: aspeed: Correct the DAT usage for set_dynamic_addr and reattach

In the Aspeed I3C HCI, the DAT uses the index as the dynamic address
rather than the [23:16] field value. Therefore, the driver needs to modify
the implementation of the following APIs:
1.alloc_entry: Needs to supply the dynamic address to obtain the correct
DAT index. This change was implemented in the previous patch:
commit 2cd10c0e40f6 ("i3c: mipi-i3c-hci: Make aspeed i3c compatible with
HCI.")
2.set_dynamic_addr: Should allocate a new entry with the dynamic address
and release the entry associated with the old dynamic address (DAT index)
if the DAT index does not match the dynamic address.
3.reattach_i3c_dev: dat_idx should be assigned to the dynamic address.

Signed-off-by: Billy Tsai <billy_tsai@aspeedtech.com>
Change-Id: I0d04f361355b41f02821402dc7a5ce0c9a56d260

ojayanth commented on drivers/i3c/master/mipi-i3c-hci/dat_v1.c:
```diff
@@ -103,7 +103,8 @@ static int hci_dat_v1_alloc_entry(struct i3c_hci *hci, unsigned int address)
 
 	/* default flags */
-	dat_w0_write(dat_idx, DAT_0_SIR_REJECT | DAT_0_MR_REJECT);
+	dat_w0_write(dat_idx,
```
Is this fix is specific to AMD CRB ?

mahkurap replied:
> Is this fix is specific to AMD CRB ?

Yes, without this change, i3ctransfer requests to sbrmi/sbtsi request fail intermittently.

mahkurap commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-nigeria.dts:
```diff
@@ -399,6 +400,23 @@
+	i3c-scl-hz = <12500000>;
+	i2c-scl-hz = <1000000>;
+
+	sbtsi_p1_0: sbtsi@4c,22400000001 {
```
correct the static slave address for sbrmi and sbtsi nodes.

mahkurap commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-morocco.dts:
```diff
@@ -619,6 +620,24 @@
+	i3c-scl-hz = <12500000>;
+	i2c-scl-hz = <1000000>;
+
+	sbtsi_p1_0: sbtsi@4c,22400000001 {
```
correct the static slave address for sbtsi/sbrmi

supven01 commented on drivers/i3c/master/mipi-i3c-hci/dat_v1.c:
```diff
@@ -141,13 +141,24 @@ static void hci_dat_v1_free_entry(struct i3c_hci *hci, unsigned int dat_idx)
+		ret = hci_dat_v1_alloc_entry(hci, address);
+		if (ret < 0)
+			dev_err(&hci->master.dev, "Allocate entry: %d", ret);
+		hci_dat_v1_free_entry(hci, dat_idx);
```
should this be freed only if there is an error? if thats the case, braces required at line 149 and and 151

supven01 commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -747,7 +747,8 @@ static int i3c_hci_attach_i3c_dev(struct i3c_dev_desc *dev)
 #ifdef CONFIG_ARCH_ASPEED
-		ret = mipi_i3c_hci_dat_v1.alloc_entry(hci, dev->info.dyn_addr);
+		ret = mipi_i3c_hci_dat_v1.alloc_entry(hci,
+						      dev->info.dyn_addr ?: dev->info.static_addr);
```
Please use full form of Ternary Condition.
(condition)? True : false;

supven01 replied:
(dev->info.dyn_addr != NULL) ? dev->info.dyn_addr : dev->info.static_addr;

mahkurap replied:
> (dev->info.dyn_addr != NULL) ? dev->info.dyn_addr : dev->info.static_addr;

fixed it. please check the new commits

supven01 commented on drivers/i3c/master/mipi-i3c-hci/core.c:
```diff
@@ -755,7 +755,8 @@ static int i3c_hci_attach_i3c_dev(struct i3c_dev_desc *dev)
 		}
-		mipi_i3c_hci_dat_v1.set_dynamic_addr(hci, ret, dev->info.dyn_addr);
+		mipi_i3c_hci_dat_v1.set_dynamic_addr(hci, ret,
+						     dev->info.dyn_addr ?: dev->info.static_addr);
```
Please use full form of Ternary Condition.
(condition)? True : false;

(dev->info.dyn_addr != NULL) ? dev->info.dyn_addr : dev->info.static_addr;

supven01 replied:
Commit ID "83c645e790e13908c55fa9acd5f1eaa867eab4ca_" has this change already. Drop one of the commits

supven01 replied:
or if it is at different place in the same core.c file. Merge them together.
